src/rhsm/utils.py

```diff
@@ -297,6 +298,104 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def singleton(cls: type) -> type:
+    """Decorate a class to make it singleton.
+
+    This decorator is inherited: subclasses will be singletons as well,
+    but they won't be the same singleton as their parent.
+    """
+
+    new = cls.__new__
+    cls._instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if not isinstance(cls._instance, cls):
+            cls._instance = new(cls)
+        return cls._instance
+
+    cls.__new__ = __new__
+    return cls
+
+
+def call_once(fn: Callable) -> Callable:
+    """Decorate a function to make it callable just once.
+
+    All further calls do not do anything and return None.
+    """
+
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        if not hasattr(fn, "_call_once_lock"):
+            fn._call_once_lock = threading.RLock()
+
+        locked: bool = fn._call_once_lock.acquire(blocking=True, timeout=1.0)
+        if not locked:
+            raise RuntimeError(f"Could not acquire call_once lock for function {fn.__name__}.")
```

Comment:
hmm wouldn't it be better to acquire the lock as a context manager? this way it will be released automatically, and we can simplify the try later on

Reply:
Hmm the documentation for it says so though,
https://docs.python.org/3.6/library/threading.html#rlock-objects
"Reentrant locks also support the context management protocol."